frame/vamm/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+		/// conversions to `u128`, which are used in the computations.
+		type Balance: Default
+			+ AtLeast32BitUnsigned
+			+ CheckedAdd
+			+ CheckedDiv
+			+ CheckedMul
+			+ CheckedSub
+			+ Codec
+			+ Copy
+			+ MaxEncodedLen
+			+ Ord
+			+ Parameter
+			+ Zero;
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                             Pallet Types
+	// ----------------------------------------------------------------------------------------------------
+
+	type BalanceOf<T> = <T as Config>::Balance;
+	type TimestampOf<T> = <T as Config>::Timestamp;
+	type VammIdOf<T> = <T as Config>::VammId;
+	type VammStateOf<T> = VammState<BalanceOf<T>, TimestampOf<T>>;
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Runtime  Storage
+	// ----------------------------------------------------------------------------------------------------
+
+	/// The number of virtual markets, also used to generate the next market
+	/// identifier.
+	///
+	/// # Note
+	///
+	/// Frozen markets do not decrement the counter.
+	#[pallet::storage]
+	#[pallet::getter(fn vamm_count)]
+	#[allow(clippy::disallowed_types)]
+	pub type VammCounter<T: Config> = StorageValue<_, VammIdOf<T>, ValueQuery>;
+
+	/// Maps [VammId](Config::VammId) to the corresponding virtual
+	/// [VammState] specs
+	#[pallet::storage]
+	#[pallet::getter(fn get_vamm)]
+	pub type VammMap<T: Config> = StorageMap<_, Blake2_128Concat, VammIdOf<T>, VammStateOf<T>>;
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                            Runtime Events
+	// ----------------------------------------------------------------------------------------------------
+
+	// Pallets use events to inform users when important changes are made.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Emitted after a successful call to the [`create`](Pallet::create) function.
+		Created { vamm_id: VammIdOf<T>, state: VammStateOf<T> },
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Runtime  Errors
+	// ----------------------------------------------------------------------------------------------------
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This error is thrown when the caller tries to set `base_asset_reserves` to zero.
+		BaseAssetReserveIsZero,
+		/// This error is thrown when the caller tries to set `quote_asset_reserves` to zero.
+		QuoteAssetReserveIsZero,
+		/// This error is thrown when the caller tries to set `peg_multiplier` to zero.
+		PegMultiplierIsZero,
+	}
```

Comment:
May be worthwhile to remove the repetitive 'prefixes', e.g., 'Caller tried to set `base_asset_reserves` to zero.

Reply:
Alright :+1: 

Made it more succinct.